File: src/main/java/com/vc/medium/L508.scala
package com.vc.medium

class TreeNode508(var _value: Int) {
  var value: Int = _value
  var left: TreeNode508 = null
  var right: TreeNode508 = null
}


object L508 {
  def findFrequentTreeSum(root: TreeNode508): Array[Int] = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, Int]()
    def solve(root: TreeNode508): Int = {
      if(root != null) {
        val l = solve(root.left)
        val r = solve(root.right)
        val total = l + r + root.value
        map.put(total, map.getOrElse(total, 0) + 1)
        return total
      }
      0
    }

    val res = new mutable.ArrayBuffer[Int]()
    solve(root)
    if(map.nonEmpty) {
      val maxValue = map.maxBy(_._2)._2
      map.foreach(x => if(x._2 == maxValue) res += x._1)
    }
    res.toArray
  }
}

File: src/main/java/com/vc/medium/L508.java
package com.vc.medium;

import com.vc.hard.TreeNode;

import java.util.*;

class L508 {
    private HashMap<Integer, Integer> map;
    public int[] findFrequentTreeSum(TreeNode root) {
        this.map = new HashMap<>();
        helper(root);

        int maxCount = 0;
        List<Integer> max = new ArrayList<>();
        for(Map.Entry<Integer, Integer> entry: map.entrySet()) {
            int sum = entry.getKey();
            int sumCount = entry.getValue();
            if(sumCount > maxCount) {
                maxCount = sumCount;
                max.clear();
                max.add(sum);
            }
            else if(sumCount == maxCount) max.add(sum);
        }

        int[] res = new int[max.size()];
        for(int i = 0; i < max.size(); i++) res[i] = max.get(i);
        return res;
    }

    private int helper(TreeNode root) {
        if(root == null) return 0;

        int left = helper(root.left);
        int right = helper(root.right);

        int sum = left + right + root.val;
        map.put(sum, map.getOrDefault(sum, 0) + 1);

        return sum;
    }
}
